README.md
# Darian Mars Calendar Converter

This is Ruby library to convert Earth time to Mars time in Darian calendar.

The [Darian calendar] is a proposed system of time-keeping on the planet Mars.
It was created by aerospace engineer and political scientist Thomas Gangale
in 1985. This library use 2002 version of calendar with Telescopic Epoch
(1609 instead of 1975 year in original calendar propose).

Based on [JS converter] by Thomas Gangale.

Sponsored by [Evil Martians].

[Darian calendar]: http://en.wikipedia.org/wiki/Darian_calendar
[JS converter]: http://pweb.jps.net/~tgangale/mars/converter/calendar_clock.htm
[Evil Martians]: http://evilmartians.com/

## Usage

```ruby
require 'darian'

earth_time = Time.now
mars_time  = Darian.from_earth(earth_time)

mars_time.to_s #=> "214-09-17 15:07:17"

mars_time.year  #=> 214
mars_time.month #=> 9
mars_time.sol   #=> 17
mars_time.day   #=> 17

mars_time.month_name    #=> "Aries"
mars_time.week_sol_name #=> "Sol Martis"

earth_date = Date.today
mars_date  = Darian.from_earth(earth_date)
```

## License

Library is licensed under the GNU Lesser General Public License version 3.
See the LICENSE file or http://www.gnu.org/licenses/lgpl.html.

## Author

Andrey “A.I.” Sitnik <andrey@sitnik.ru>


ChangeLog
== 0.1 (Viking 2)
* Add Date.today and Time.now shortcuts.
* Add parse_earth shortcut to Date and Time.
* Allow to compare dates and times.

== 0.0.3 (Viking 1)
* Convert DateTime to martian time.
* Fix documentation.

== 0.0.2 (Mars 6)
* Add Date class and converter.

== 0.0.1 (Mars 3)
* Initial release


spec/date_spec.rb
# encoding: utf-8
require File.expand_path('../spec_helper', __FILE__)

describe Darian::Date do

  before do
    earth = Date.parse('2012-05-16')
    @mars = Darian::Date.from_earth(earth)
  end

  it "should convert Earth time to Mars" do
    @mars.year.should  == 214
    @mars.month.should == 9
    @mars.sol.should   == 17

    @mars.since_epoch.should == 143318
    @mars.week_sol.should    == 3

    @mars.season.should          == 1
    @mars.month_of_season.should == 2
  end

  it "should set month and week sol name" do
    @mars.month_name.should    == 'Aries'
    @mars.week_sol_name.should == 'Sol Martis'
  end

  it "should alias day to sol" do
    @mars.day.should           == @mars.sol
    @mars.week_day.should      == @mars.week_sol
    @mars.week_day_name.should == @mars.week_sol_name
  end

  it "should print mars time" do
    @mars.to_s.should == '214-09-17'
  end

  it "should parse Earth date" do
    Darian::Date.parse_earth('2012-05-16').should == @mars
  end

  it "should compare dates" do
    past   = Darian::Date.parse_earth('2012-05-15')
    same   = Darian::Date.parse_earth('2012-05-16')
    future = Darian::Date.parse_earth('2012-05-17')

    @mars.should >  past
    @mars.should <  future
    @mars.should == same

    @mars.should_not < past
    @mars.should_not > future
  end

end


spec/time_spec.rb
# encoding: utf-8
require File.expand_path('../spec_helper', __FILE__)

describe Darian::Time do

  before do
    earth = Time.parse('2012-05-16 10:00:00 UTC')
    @mars = Darian::Time.from_earth(earth)
  end

  it "should convert Earth time to Mars" do
    @mars.year.should  == 214
    @mars.month.should == 9
    @mars.sol.should   == 17
    @mars.hour.should  == 15
    @mars.min.should   == 7
    @mars.sec.should   == 17

    @mars.since_epoch.should == 143318.630064585
    @mars.week_sol.should    == 3

    @mars.season.should          == 1
    @mars.month_of_season.should == 2
  end

  it "should set month and week sol name" do
    @mars.month_name.should    == 'Aries'
    @mars.week_sol_name.should == 'Sol Martis'
  end

  it "should alias day to sol" do
    @mars.day.should           == @mars.sol
    @mars.week_day.should      == @mars.week_sol
    @mars.week_day_name.should == @mars.week_sol_name
  end

  it "should print mars time" do
    @mars.to_s.should == '214-09-17 15:07:17'
  end

  it "should convert to date" do
    date = @mars.to_date
    date.should be_a(Darian::Date)
    date.to_s.should == '214-09-17'
  end

  it "should parse Earth time" do
    str = '2012-05-16 10:00:00 UTC'
    Darian::Time.parse_earth(str).should == @mars
  end

  it "should compare times" do
    past   = Darian::Time.parse_earth('2012-05-16 09:59:59 UTC')
    same   = Darian::Time.parse_earth('2012-05-16 10:00:00 UTC')
    future = Darian::Time.parse_earth('2012-05-16 10:00:01 UTC')

    @mars.should >  past
    @mars.should <  future
    @mars.should == same

    @mars.should_not < past
    @mars.should_not > future
  end

end


spec/darian_spec.rb
# encoding: utf-8
require File.expand_path('../spec_helper', __FILE__)

describe Darian do

  it "should return own version" do
    Darian::VERSION.should be_a(String)
  end

  it "should convert by argument class" do
    Darian.from_earth(Time.now).should     be_a(Darian::Time)
    Darian.from_earth(DateTime.now).should be_a(Darian::Time)
    Darian.from_earth(Date.today).should   be_a(Darian::Date)
    lambda { Darian.from_earth(1) }.should raise_error(ArgumentError)
  end

  it "should return current time" do
    earth = Time.parse('2012-05-16 10:00:00 UTC')
    Time.should_receive(:now).and_return(earth)

    Darian.now.should == Darian::Time.from_earth(earth)
  end

  it "should return current date" do
    earth = Date.parse('2012-05-16')
    Date.should_receive(:today).and_return(earth)

    Darian.today.should == Darian::Date.from_earth(earth)
  end

end


lib/darian.rb
# encoding: utf-8
=begin
Main file to load all neccessary classes for Darian Mars calendar.

Copyright (C) 2012 Andrey “A.I.” Sitnik <andrey@sitnik.ru>,
sponsored by Evil Martians.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
=end

dir = File.join(File.dirname(__FILE__), 'darian')
require File.join(dir, 'version')
require File.join(dir, 'date_methods')
require File.join(dir, 'time')
require File.join(dir, 'date')

# Darian Mars calendar.
#
#   mars_time = Darian.from_earth(earth_time)
module Darian
  class << self

    # Convert `arg` to martian time or date, depend on `arg` class.
    def from_earth(arg)
      if arg.is_a? ::Time
        Darian::Time.from_earth(arg)
      elsif arg.is_a? ::DateTime
        Darian::Time.from_earth(::Time.parse(arg.to_s))
      elsif arg.is_a? ::Date
        Darian::Date.from_earth(arg)
      else
        raise ArgumentError, "Can't convert #{arg.class} to martian time or date"
      end
    end

    # Return current Mars time. Shortcut for `Darian.from_earth(Time.now)`.
    def now
      Darian::Time.from_earth(::Time.now)
    end

    # Return current Mars date. Shortcut for `Darian.from_earth(Date.today)`.
    def today
      Darian::Date.from_earth(::Date.today)
    end

  end
end


lib/darian/date.rb
# encoding: utf-8
=begin
Darian Mars calendar date converter.

Copyright (C) 2012 Andrey “A.I.” Sitnik <andrey@sitnik.ru>,
sponsored by Evil Martians.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
=end

module Darian
  # Darian Mars calendar date converter.
  #
  #   mars = Darian::Date.from_earth(Date.today)
  #   puts mars
  #   puts mars.month_name
  class Date
    include DateMethods

    # Return Mars date converted from Earth date.
    #
    #   Darian.from_earth(Date.today)
    def self.from_earth(date)
      time = ::Time.parse(date.to_s + " 00:00:00 UTC")
      Darian::Time.from_earth(time).to_date
    end

    # Parse Earth date and convert to Mars date.
    # Shortcut for `Darian::Date.from_earth(Date.parse(string))`.
    def self.parse_earth(string)
      self.from_earth(::Date.parse(string))
    end

    # Create martian date by martian time.
    def initialize(time)
      @since_epoch = time.since_epoch.floor

      @year     = time.year
      @month    = time.month
      @sol      = time.sol
      @week_sol = time.week_sol

      @season          = time.season
      @sol_of_season   = time.sol_of_season
      @month_of_season = time.month_of_season
    end

    # Printable string of martian date.
    def to_s
      sprintf '%d-%02d-%02d', @year, @month, @sol
    end
  end
end


lib/darian/time.rb
# encoding: utf-8
=begin
Darian Mars calendar time converter.

Copyright (C) 2012 Andrey “A.I.” Sitnik <andrey@sitnik.ru>,
sponsored by Evil Martians.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
=end

module Darian
  # Darian Mars calendar time converter.
  #
  #   mars = Darian::Time.from_earth(Time.now)
  #   puts mars
  #   puts mars.month_name
  class Time
    MARS_TO_EARTH_DAYS = 1.027491251
    EPOCH_OFFSET       = 587744.77817
    ROUND_UP_SECOND    = 1/86400

    include DateMethods

    attr_reader :hour
    attr_reader :min
    attr_reader :sec

    # Return Mars time converted from Earth time.
    #
    #   Darian.from_earth(Time.now)
    def self.from_earth(time)
      # 719527 is days until 0 year for Unix Epoch
      days = (time.to_i / 86400.0) + 719527
      sols = (days - EPOCH_OFFSET) / MARS_TO_EARTH_DAYS
      self.new(sols)
    end

    # Parse Earth time and convert to Mars time.
    # Shortcut for `Darian::Time.from_earth(Time.parse(string))`.
    def self.parse_earth(string)
      self.from_earth(::Time.parse(string))
    end

    # Create Mars time by sols since 0 year.
    #
    # It is internal contructor. Use `Darian.from_earth`.
    def initialize(sols)
      @since_epoch = sols

      sD  = (sols / 334296).floor
      doD = (sols - (sD * 334296)).floor

      sC   = 0
      doC  = doD
      sC   = ((doD - 1) / 66859).floor if doD != 0
      doC -= (sC * 66859) + 1 if sC != 0

      sX  = 0
      doX = doC
      if sC != 0 # century that does not begin with leap day
        sX   = ((doC + 1) / 6686).floor
        doX -= sX * 6686 - 1 if sX != 0
      else
        sX   = (doC / 6686).floor
        doX -= sX * 6686 if sX != 0
      end

      sII = 0
      doII = doX
      if sC != 0 and sX == 0 # decade that does not begin with leap day
        sII   = (doX / 1337).floor
        doII -= sII * 1337 if sII != 0
      else  # 1338, 1337, 1337, 1337 …
        sII   = ((doX - 1) / 1337).floor if doX != 0
        doII -= (sII * 1337 + 1) if sII != 0
      end

      sI  = 0
      doI = doII
      if sII == 0 and (sX != 0 or (sX == 0 and sC == 0))
        sI   = (doII / 669).floor
        doI -= 669 if sI != 0
      else # 668, 669
        sI   = ((doII + 1) / 669).floor
        doI -= 668 if sI != 0
      end

      @year = 500 * sD + 100 * sC + 10 * sX + 2 * sII + sI
      @season = if doI < 167
        0
      elsif doI < 334
        1
      elsif doI < 501
        2
      else
        3
      end
      @sol_of_season   = doI - 167 * @season
      @month_of_season = (@sol_of_season / 28).floor

      @month = @month_of_season + 6 * @season + 1

      @sol      = doI - ((@month - 1) * 28 - @season) + 1
      @week_sol = (@sol - 1) % 7 + 1

      hour = (sols - sols.floor) * 24
      min  = (hour - hour.floor) * 60

      @hour = hour.floor
      @min  = min.floor
      @sec  = ((min - min.floor) * 60).floor
    end

    # Printable string of martian time.
    def to_s
      sprintf '%d-%02d-%02d %02d:%02d:%02d',
              @year, @month, @sol, @hour, @min, @sec
    end

    # Return martian date.
    def to_date
      Darian::Date.new(self)
    end
  end
end


lib/darian/date_methods.rb
# encoding: utf-8
=begin
Common date methods for Time and Date classes.

Copyright (C) 2012 Andrey “A.I.” Sitnik <andrey@sitnik.ru>,
sponsored by Evil Martians.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
=end

module Darian
  # Common date methods for Time and Date classes.
  module DateMethods
    include Comparable

    attr_reader :year
    attr_reader :month
    attr_reader :sol

    attr_reader :since_epoch
    attr_reader :week_sol
    attr_reader :season
    attr_reader :sol_of_season
    attr_reader :month_of_season

    alias :day      :sol
    alias :week_day :week_sol

    def month_name
      case @month
        when  1 then 'Sagittarius'
        when  2 then 'Dhanus'
        when  3 then 'Capricornus'
        when  4 then 'Makara'
        when  5 then 'Aquarius'
        when  6 then 'Kumbha'
        when  7 then 'Pisces'
        when  8 then 'Mina'
        when  9 then 'Aries'
        when 10 then 'Mesha'
        when 11 then 'Taurus'
        when 12 then 'Rishabha'
        when 13 then 'Gemini'
        when 14 then 'Mithuna'
        when 15 then 'Cancer'
        when 16 then 'Karka'
        when 17 then 'Leo'
        when 18 then 'Simha'
        when 19 then 'Virgo'
        when 20 then 'Kanya'
        when 21 then 'Libra'
        when 22 then 'Tula'
        when 23 then 'Scorpius'
        when 24 then 'Vrishika'
      end
    end

    def week_sol_name
      case @week_sol
        when 1 then 'Sol Solis'
      	when 2 then 'Sol Lunae'
      	when 3 then 'Sol Martis'
      	when 4 then 'Sol Mercurii'
      	when 5 then 'Sol Jovis'
      	when 6 then 'Sol Veneris'
      	when 7 then 'Sol Saturni'
      end
    end
    alias :week_day_name :week_sol_name

    def <=>(another)
      since_epoch <=> another.since_epoch
    end
  end
end
